Pack runtime arguments across brisc/ncrisc/trisc

pgkeller opened the pull request:
### Ticket
#8602 
### Problem description
Runtime args are sent w/ many unicast NOC commands.  Packing B/N/T together reduces the number of commands by up to a factor of 3 which improves perf when few RTAs are sent

### What's changed
Track number of RTAs in kernel groups
"Level" (to max) the number of RTAs per dispatch processor type (B/N/T)
Send RTAs w/ single packed write command per core (for unique RTAs)

### Checklist
- [ ] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

pgkeller committed:
#8602: Downgrade RISCV from enum class to enum

Want to use this in too many arrays, need access to ordinal values

pgkeller committed:
#8602: More runtime_args test tweaks

To support dynamic mem map

pgkeller committed:
#8602: Make get_core_range_set return a const reference

Was returning a temporary leading to crashes when used in an iterator

pgkeller committed:
#8602: Pack runtime args in memory map

B/N/T RTAs are densely packed
Unique RTAs are sent w/ 1 packed-write command
Common RTAs are sent w/ 1 packed-write command (TODO: use linear writes)

pgkeller committed:
#0: Comment out some prefetcher dprints

pgkeller committed:
#8602: Update multi-core-range runtime args tests

Adds a test w/ a range doesn't get merged into one kernel group
Does more arg updating

tt-aho commented on tests/tt_metal/tt_metal/unit_tests_fast_dispatch/command_queue/test_EnqueueProgram.cpp:
```diff
@@ -1265,8 +1332,8 @@ TEST_F(CommandQueueFixture, TestRandomizedProgram) {
         std::shuffle(std::begin(programs), std::end(programs), rng);
-        if (i % 10 == 0) {
-            log_debug(tt::LogTest, "Enqueueing {} programs for iter: {}/{} now.", programs.size(), i+1, NUM_ITERATIONS);
+        if (i % 50 == 0) {
+            log_info(tt::LogTest, "Enqueueing {} programs for iter: {}/{} now.", programs.size(), i+1, NUM_ITERATIONS);
```
Do we want this to actually be log_info, or revert to log_debug?

pgkeller replied:
I like this as log_info, otherwise when I'm running tests I keep wondering if the test has hung

tt-aho commented on tt_metal/hw/firmware/src/ncrisc.cc:
```diff
@@ -15,6 +15,7 @@
 #include "circular_buffer.h"
 
 #include "debug/status.h"
+#include "debug/dprint.h"
```
Remove?

tt-aho commented on tt_metal/hw/firmware/src/trisc.cc:
```diff
@@ -13,6 +13,7 @@
 
 #include "debug/fw_debug.h"
 #include "debug/status.h"
+#include "debug/dprint.h"
```
Remove?

tt-aho commented on tt_metal/impl/program/program.cpp:
```diff
@@ -768,6 +759,103 @@ void Program::populate_dispatch_data(Device *device) {
+
+        // TODO: this is asserted here as the leveling above can break the limits enforced by the API
+        // Once we use a ring buffer, memory space will be dynamic and this assert won't matter
+        TT_ASSERT(offset <= L1_KERNEL_CONFIG_SIZE);
```
Should this be fatal instead of assert?

tt-aho reviewed (approved):
overall lgtm. Minor/trivial comments/questions

aliuTT commented on tt_metal/impl/program/program.hpp:
```diff
@@ -131,6 +134,10 @@ class Program {
 
+    void finalize_rt_args();
+    bool is_finalized() const { return loaded_onto_device; }
+    void set_finalized() { loaded_onto_device = true; }
```
Can you change `program.loaded_onto_device = true;` from `command_queue.cpp` around line 1200 to use this api?

pgkeller replied:
will do

pgkeller commented on tests/tt_metal/tt_metal/unit_tests/basic/runtime_args.cpp:
```diff
@@ -39,7 +39,9 @@ uint32_t get_runtime_arg_addr(tt::RISCV processor) {
     }
-    return result_base;
+
+    uint32_t offset = is_common ? 3 * 256 * sizeof(uint32_t) : 0;
```
these are just in place of a memory allocator and are self contained.  I'll change them to use constexpr so it is more clear what the magic is.  eventually, I'd love for our test infra to use an allocator rather than magic addresses above the base

pgkeller committed:
#8602: Reduce size of launch message

Exceeding 32 bytes causes hangs, presumably due to PCIE transation getting
re-ordered.  Reduce size for now (also a reasonable change), will investigate
later

pgkeller committed:
#8602: Add static_assert for launch message size

Remove runtime assert
